accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+                                                         values are sha256, sha384, and sha512. [default: SHA256]
+  -m, --max-concurrency <max-concurrency>                Maximum concurrency. [default: 4]
+  -v, --verbosity                                        Sets the verbosity level. Allowed values are 'none',
+  <Critical|Debug|Error|Information|None|Trace|Warning>  'critical', 'error', 'warning', 'information', 'debug', and
+                                                         'trace'. [default: Warning]
+  -?, -h, --help                                         Show help and usage information
+```
+
+The new command can then be used to sign artifacts.
+
+```text
+sign code azure-key-vault -kvu https://my.vault.azure.test/ -kvc MyCertificate -kvm -d Description -u http://description.test -b C:\ClassLibrary1\ClassLibrary1\bin\Debug\net8.0 ClassLibrary1.dll
+```
+
+### Sign artifacts using Windows certificate store
+
+First, the plugin must be installed.  The following command would download and install the latest version of the plugin.
+
+```text
+sign plugin install Microsoft.Windows.CertificateStore.Sign
+```
+
+The new command can then be used to sign artifacts.
+
+```text
+sign code certificate-store --store-location CurrentUser --store-name My --sha1fingerprint da39a3ee5e6b4b0d3255bfef95601890afd80709 -d Description -u http://description.test -b C:\ClassLibrary1\ClassLibrary1\bin\Debug\net8.0 ClassLibrary1.dll
+```
+
+## Requirements
+
+### Goals
+
+* Create a plugin model that enables pluggable signature providers.  A signature provider plugin will offer an alternate implementation of [`System.Security.Cryptography.AsymmetricAlgorithm`](https://learn.microsoft.com/dotnet/api/system.security.cryptography.asymmetricalgorithm) and the [`System.Security.Cryptography.X509Certificates.X509Certificate2`](https://learn.microsoft.com/dotnet/api/system.security.cryptography.x509certificates.x509certificate2), if available, corresponding to the private key.  Note that while there may be valid scenarios for signing with a raw asymmetric key pair, some signing operations require a certificate and will fail if a certificate is not available.
+* Make Sign CLI plugin-neutral.  While Sign CLI may install some core plugins (TBD), most plugins should be installed separately from Sign CLI itself, and Sign CLI's only interactions with any plugin should be through this plugin model.
+* Enable Sign CLI and plugins to version and release independently.
+
+### Non-Goals
+
+* Enable signature formats that require a certificate to sign without a certificate.
+* Enable signature formats to support asymmetric algorithms they do not already support.
+* Create a distribution channel for plugins.  Sign CLI is a .NET tool and is [available on https://nuget.org](https://www.nuget.org/packages/sign/).  Plugin packages can be published to any NuGet feed, including <https://nuget.org>.
```

Comment:
If interaction with arbitrary NuGet feeds is expected to be supported shouldn't the spec cover how user would tell sign CLI which feed to use? What about authentication to feeds that require one? V2/V3 support?

Reply:
I'm also wondering if NuGet-based plugins should be developed as a separate feature. Seems useful.

Reply:
One other consideration - the current description seems to hint at the nuget package being just downloaded as a zip file and extracted onto disk - no real NuGet logic involved. This will mean that the host will have to implement some of the logic NuGet client already implements (like TFM resolution). Would it make more sense to resolve the package via nuget client instead? That would provide all of the resolution logic as well as authentication and so on.

Reply:
@agr, @vitek-karas, good questions.  We should use NuGet API's.  This would enable TFM resolution, support alternate package feeds, authentication, etc.  The spec should detail this.

@richlander, I agree a general model does seem like it might be useful.  @clairernovotny and I were hard-pressed to find close precedents here, so what we're doing seemed novel.  We're open to working with others on a generalized model if it would be useful.